packages/aztec.js/src/proof/joinSplitFluid/index.js

```diff
@@ -0,0 +1,106 @@
+const { constants, errors, proofs } = require('@aztec/dev-utils');
+const { AbiCoder } = require('web3-eth-abi');
+const { keccak256 } = require('web3-utils');
+
+const { inputCoder, outputCoder } = require('../../encoder');
+const JoinSplitProof = require('../joinSplit');
+const { ProofType } = require('../proof');
+const ProofUtils = require('../utils');
+
+const { AztecError } = errors;
+
+class JoinSplitFluidProof extends JoinSplitProof {
+    constructor(
+        type,
+        currentTotalValueNote,
+        newTotalValueNote,
+        adjustedNotes,
+        sender,
+        metadata = [currentTotalValueNote, ...adjustedNotes],
+    ) {
+        const publicValue = constants.ZERO_BN;
+        const publicOwner = constants.addresses.ZERO_ADDRESS;
+        super([newTotalValueNote], [currentTotalValueNote, ...adjustedNotes], sender, publicValue, publicOwner, metadata);
+
+        // Overriding this because JoinSplit sets the type to `JOIN_SPLIT`
+        if (type !== ProofType.BURN.name && type !== ProofType.MINT.name) {
+            throw new Error(`proof type should be one of ${[ProofType.BURN.name, ProofType.MINT.name]}`);
+        }
+        this.type = type;
+    }
+
+    /**
+     * We may remove this function in the future and let the upper JoinSplitProof class handle
+     * the challenge construction. In the interim, we need it because the join split fluid
+     * validator doesn't expect the public owner in the challenge.
+     */
+    constructChallenge() {
+        this.constructChallengeRecurse([this.sender, this.publicValue, this.m, this.notes, this.blindingFactors]);
+        this.challenge = this.challengeHash.redKeccak();
+    }
+
+    constructOutputs() {
+        this.output = '';
+        this.outputs = outputCoder.encodeProofOutputs([
+            {
+                inputNotes: [
+                    {
+                        ...this.outputNotes[0],
+                        forceMetadata: true,
+                    },
+                ],
+                outputNotes: [
+                    {
+                        ...this.inputNotes[0],
+                        forceNoMetadata: true,
+                    },
+                ],
+                publicValue: this.publicValue,
+                publicOwner: this.publicOwner,
+                challenge: this.challengeHex,
+            },
+            {
+                inputNotes: [],
+                outputNotes: this.outputNotes.slice(1),
+                publicValue: this.publicValue,
+                publicOwner: this.publicOwner,
+                challenge: keccak256(this.challengeHex), // TODO: figure out why this is the challenge hex and not the note the challenge itself
```

Comment:
We can remote this comment, `web3.keccak256` will treat the input as hex-encoded, if the string starts with `0x`